cpp/include/games/nim/Game.hpp

```diff
@@ -0,0 +1,157 @@
+    static constexpr int kNumPlayers = nim::kNumPlayers;
+    static constexpr int kNumActions = nim::kMaxStonesToTake;
+    static constexpr int kMaxBranchingFactor = nim::kMaxStonesToTake;
+    static constexpr int kNumPreviousStatesToEncode = 0;
+    static constexpr float kOpeningLength = 0.1;  // not applicable to Nim
+  };
+
+  struct State {
+    auto operator<=>(const State& other) const = default;
+
+    size_t hash() const {
+      auto tuple = std::make_tuple(stones_left, current_player);
+      std::hash<decltype(tuple)> hasher;
+      return hasher(tuple);
+    }
+
+    int stones_left;
+    int current_player;
+  };
+
+  using GameResults = core::WinShareResults<Constants::kNumPlayers>;
+  using StateHistory = core::SimpleStateHistory<State, Constants::kNumPreviousStatesToEncode>;
+  using SymmetryGroup = groups::TrivialGroup;
+  using Symmetries = core::TrivialSymmetries;
+  using Types = core::GameTypes<Constants, State, GameResults, SymmetryGroup>;
+
+  struct Rules {
+    static void init_state(State& state) {
+      state.stones_left = nim::kStartingStones;
+      state.current_player = 0;
+    }
+
+    static Types::ActionMask get_legal_moves(const StateHistory& history) {
+      const State& state = history.current();
+      Types::ActionMask mask;
+
+      for (int i = 0; i < nim::kMaxStonesToTake; ++i) {
+        mask[i] = i + 1 <= state.stones_left;
+      }
+
+      return mask;
+    }
+
+    static core::seat_index_t get_current_player(const State& state) {
+      return state.current_player;
+    }
+
+    static void apply(StateHistory& history, core::action_t action) {
+      if (action < 0 || action >= nim::kMaxStonesToTake) {
+        throw std::invalid_argument("Invalid action: " + std::to_string(action));
+      }
+
+      State& state = history.extend();
+      state.stones_left -= action + 1;
+      state.current_player = 1 - state.current_player;
+    }
+
+    static bool is_terminal(const State& state, core::seat_index_t last_player,
+                            core::action_t last_action, GameResults::Tensor& outcome) {
+      if (state.stones_left == 0) {
+        outcome.setZero();
+        outcome(last_player) = 1;
+        return true;
+      }
+      return false;
+    }
+  };
+
+  struct IO {
+    static std::string action_delimiter() { return "-"; }
+    static std::string action_to_str(core::action_t action) { return std::to_string(action); }
+    static void print_state(std::ostream&, const State&, core::action_t last_action = -1,
+                            const Types::player_name_array_t* player_names = nullptr) {
+                              throw std::runtime_error("Not implemented");
+                            }
+    static void print_mcts_results(std::ostream&, const Types::PolicyTensor& action_policy,
+                                   const Types::SearchResults&) {
+                                      throw std::runtime_error("Not implemented");
+                                   }
```

Comment:
Indentation is a bit funny with these 2 functions. VSCode might do it this way by default, but this is a bad default behavior in my opinion.